File: 253-meeting-rooms-ii/meeting-rooms-ii.java
class Solution {
    public int minMeetingRooms(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        int maxSize = 0;

        for (int[] interval : intervals) {
            int startTime = interval[0];
            int endTime = interval[1];
            while (!pq.isEmpty() && pq.peek() <= startTime) {
                pq.poll();
            }
            pq.offer(endTime);
            maxSize = Math.max(pq.size(), maxSize);
        }
        return maxSize;
    }
}

File: 253-meeting-rooms-ii/meeting-rooms-ii.py
class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        start = sorted([i[0] for i in intervals])
        end = sorted(i[1] for i in intervals)

        p1 = 0
        p2 = 0
        counter = 0
        maxi = 0
        while p1 < len(intervals):
            if end[p2] <= start[p1]:
                counter -= 1
                p2 += 1
            else:
                counter += 1
                p1 += 1
            maxi = max(maxi, counter)
        return maxi

